App/global style/smooth transition

patrycja-gomza opened the pull request:
## Implement AnimatedPage and Enhancements in Header and MovieList Components

Hi Team 👋,

This pull request introduces several enhancements to the application:


### AnimatedPage Component:
Added a new AnimatedPage component that leverages the react-spring library for smooth page transitions.
Integrated the AnimatedPage component into various sections of the application, providing a visually appealing transition effect.


### Header Component Updates:
Added a subtle underline effect on hover for navigation links in the Header component.


### MovieList Component:
Implemented a loading indicator during the movie list fetching process. Dispatched the setLoading action at the start of the fetchMovieListHandler. Added a delay of 1000 milliseconds to provide a smoother user experience.
Cleared the loading state in the finally block, ensuring the indicator is removed even in case of errors.

Please review 🚀 Thank you! 😊

bedicooper commented on src/features/movies/MovieList/movieListSaga.js:
```diff
@@ -7,13 +7,16 @@ import {
 
 function* fetchMovieListHandler() {
     try {
+        yield put(setLoading(true));
```
We don't need to dispatch this action:
1. I used this method in one of our previous assigments and was informed that this was a mistake. we shouldn't have reducers that only switch state this way. It's better to add the state change to another action. Moreover `loading: true` is in initialState, so we don't need this command
2. We decided to create a state property one called `status` and give it one of those values: 'loading' 'error' 'success'

patrycja-gomza replied:
Thank you for pointing that out! 🙏 I appreciate your input. I wasn't aware that having reducers that directly switch state might not be the best practice. No one had brought this to my attention before. I will make the necessary corrections and apply the updated approach. Thanks for bringing this to my notice and sharing your knowledge! 🚀

bedicooper commented on src/features/movies/MovieList/movieListSaga.js:
```diff
@@ -7,13 +7,16 @@ import {
 function* fetchMovieListHandler() {
     try {
+        yield put(setLoading(true));
+        yield delay(1000);
```
This i faaar to long 😄 if we must 200ms tops, please 😉 But really we do not need that delay at all. Page change will animate regardles, sometimes showing the spinner and sometimes not. Why forcefully delay the transition?

patrycja-gomza replied:
Thank you for your comment above; you've convinced me. I've changed my mind and acknowledge that you're right; we shouldn't artificially delay the page transition. I will make the necessary adjustments and implement the updated approach. Thanks! 😊

bedicooper commented on src/features/movies/MovieList/movieListSaga.js:
```diff
@@ -26,6 +29,9 @@ function* fetchMovieListHandler() {
         yield call(alert, "Error fetching popular movies!");
     }
+    finally {
+        yield put(setLoading(false));
```
like in previous comment, I think we can omit this part if we update setError and setMovieList with also updating `status` property of the state.

patrycja-gomza replied:
Noted! I will correct this 😉

bedicooper reviewed (commented):
Wow! Awsome work! react-spirng library looks fantastic! I appreciate your relentless and innovative approach to elevate user experience to a new level. These transitions definitely improve the experience of using our application, bravo!

patrycja-gomza commented on src/Navigation/Header/styled.js:
```diff
@@ -78,12 +78,29 @@ export const StyledNavLink = styled(NavLink)`
     padding: 0px 24px;
-    border-radius: 24px;
-    transition: 0.3s;
+    position: relative;
     color: ${({ theme }) => theme.color.white};
+    border: 1px solid transparent;
+
+    &::after {
+        content: '';
+        position: absolute;
+        width: 50%;
+        height: 0.5px;
+        bottom: -1px;
+        left: 25%;
+        background-color: transparent;
+        transition: background-color 0.6s ease-in-out;
+    }
+
+    &:hover::after {
+        background-color: ${({ theme }) => theme.color.white};
+        bottom: -1px;
+    }
 
     &.active{
         border: 1px solid ${({ theme }) => theme.color.white};
+        border-radius: 24px;
```
@bedicooper, thank you for your opinion 🙌 
I agree with you that my code is too long and too complicated. I appreciate the suggestion to simplify the code, and I am willing to implement it 🎉 
Regarding the hover effect, I considered your suggestion and tested a few other possibilities. I would like to propose a subtle brightening of the text when the cursor hovers over it:
`&:hover {`
`background:` 
`radial-gradient(ellipse farthest-side at center, rgba(255, 255, 255, 0.3) 0%, rgba(255, 255, 255, 0.2) 50%, rgba(255, 255, 255, 0) 100%);`
`}`
Let me know what you think about this idea 🙂

bedicooper replied:
It's certainly an interesting effect, I even like it, but can it be turned off when hovering over an active link?

patrycja-gomza commented on src/common/AnimatedPage.js:
```diff
@@ -0,0 +1,25 @@
+import { useHistory } from 'react-router-dom';
+
+const AnimatedPage = ({ children }) => {
+    const history = useHistory();
```
@bedicooper indeed the history constant is not used anywhere and was left behind after my earlier tests. I will remove it. Thanks for pointing it out 🙏

bedicooper replied:
I leave stuff behind all the time, one would think I should read my own code more often before commiting 😉

patrycja-gomza commented on src/Navigation/Header/styled.js:
```diff
@@ -78,12 +78,29 @@ export const StyledNavLink = styled(NavLink)`
+
+    &:hover::after {
+        background-color: ${({ theme }) => theme.color.white};
+        bottom: -1px;
```
Thank you for your feedback! I have now implemented a completely different solution for the hover state. I hope you find it satisfactory 😊

patrycja-gomza commented:
> Wow! Awsome work! react-spirng library looks fantastic! I appreciate your relentless and innovative approach to elevate user experience to a new level. These transitions definitely improve the experience of using our application, bravo!
> 
> As allways I have some small comments and suggestions (I'm sorry, it's a disease) so feel free to use them or not 💪

Hi @bedicooper 👋
Thank you so much for your encouraging words! 🥰 I'm glad you appreciate the use of the react-spring library and the effort to enhance the user experience. I've made several adjustments to the code based on your review. I always welcome feedback, so please feel free to share your comments and suggestions. Your input is valuable, and together we can continue to refine and improve the application! 💪

bedicooper commented on src/common/AnimatedPage.js:
```diff
@@ -0,0 +1,22 @@
+        opacity: 1,
+        from: { opacity: 0 },
+        config: {
+            duration: 800,
```
I forgot to comment on that in previous review, please check if shorter duration ie. 400ms doesn't look better (it does to me)

patrycja-gomza replied:
I've tested the shorter duration of 400ms for the opacity animation, and I agree with you; it does look better and adds a quicker, smoother transition. I've made the change as per your recommendation. 
Thanks for the suggestion! 😊

bedicooper commented on src/features/movies/MovieList/movieListSaga.js:
```diff
@@ -1,4 +1,4 @@
-import { all, call, put, takeEvery, select } from "redux-saga/effects";
+import { all, call, put, takeEvery, select, delay } from "redux-saga/effects";
```
we can remove the delay import now 😉

patrycja-gomza replied:
Good shot! I appreciate your sharp eye. The delay import is indeed unnecessary. I'll remove it 🙂

bedicooper reviewed (commented):
Thanks for the changes, few final things to address in the comments and we will be ready to launch it 🚀 😄

patrycja-gomza commented on src/Navigation/Header/styled.js:
```diff
@@ -78,14 +78,26 @@ export const StyledNavLink = styled(NavLink)`
         border: 1px solid ${({ theme }) => theme.color.white};
     };
 
+    @media (max-width: ${({ theme }) => theme.breakpoints.mobileMax}px){
```
Good catch! Thanks for pointing that out. You're absolutely right; it seems there's a duplication in the code. The border-radius property should be added to the @media block on line 101, not in the &.active block. Thanks for your attention to detail! 🙂

patrycja-gomza commented on src/Navigation/Header/styled.js:
```diff
@@ -78,14 +78,26 @@ export const StyledNavLink = styled(NavLink)`
+    transition: 0.2s ease-in-out;
+
+    &:hover {
+        background: radial-gradient(ellipse farthest-side at center, rgba(255, 255, 255, 0.3) 0%, rgba(255, 255, 255, 0.2) 50%, rgba(255, 255, 255, 0) 100%);
```
Great observation! Adding background: none; to the &.active selector is a good idea to disable the hover effect on active links. I have tested this on mobiles and there is no need to repeat this code.
Thank you for being so careful with the details! 🙏✨

patrycja-gomza commented:
Hi @bedicooper 👋

I addressed the suggestions from the latest review by removing the unnecessary 'delay' import in the saga, adjusting the animation duration in the AnimatedPage component and update responsive styles in StyledNavLink.

I hope the adjustments meet the team's expectations. I would greatly appreciate your approval 😊

bedicooper reviewed (approved):
Fantastic, all looks great to me. 
🤞 🤞 🤞 Fingers crossed that merging with new library won't brake anything. 
👍 Good luck in resolving conflicts that arose, they looks like a heck of careful manual work 😉
